### ISSUE TITLE
Failed services on Steam Deck

### ISSUE THREAD

**Issue by docwhat** [ValveSoftware/SteamOS#840]
#### Your system information

* Steam client version: 1661822214
* SteamOS version: 3.3.1
* Opted into Steam client beta?: Yes (Steam Deck with Preview enabled)
* Opted into SteamOS beta?: Yes (Steam Deck with Preview enabled) 
* Have you checked for updates in Settings > System?: Yes

#### Please describe your issue in as much detail as possible:

No services should be failed.  I "Switched to Desktop Mode" and I get these failed services:

```sh
(deck@steamdeck ~)$ systemctl --failed
  UNIT                                  LOAD   ACTIVE SUB    DESCRIPTION                                 
● grub-recordfail.service               loaded failed failed Record successful boot for GRUB
● shadow.service                        loaded failed failed Verify integrity of password and group files
● steamos-finish-oobe-migration.service loaded failed failed Finish migration from OOBE build

LOAD   = Reflects whether the unit definition was properly loaded.
ACTIVE = The high-level unit activation state, i.e. generalization of SUB.
SUB    = The low-level unit activation state, values depend on unit type.
3 loaded units listed.
(deck@steamdeck ~)$
```

#### Steps for reproducing this issue:

1. Switch to Desktop Mode (Steam Deck) 
2. `systemctl --failed`

#### Notes by docwhat

- `grub-recordfail.service` is failing because it is trying to write to `/boot/grub/grubenv`.   It is trying to unset any `recordfail` being set by `grub-editenv`.  Since the partition is readonly, this isn't a problem.  You can `systemctl disable --now grub-recordfail.service` to prevent this error in the future.
- `shadow.service` is failing because the directory `/var/lib/ntp` is missing.  It's checking `/etc/passwd` and is making sure all the home directories exist.  Either remove ntpd or created the directory.
- `steamos-finish-oobe-migration.service` is incorrectly failing.  The `bash` script `/usr/lib/steamos/steamos-finish-oobe-migration` has this line:

    ```bash
    # Shouldn't have been invoked if $old_home isn't here or is a symlink.
    [[ -d $old_home && ! -L $old_home ]] || die "'$old_home' does not exist or is not a folder, nothing to migrate"
    ```
    
    It should read:
    
    ```bash
    # Shouldn't have been invoked if $old_home isn't here or is a symlink.
    if [[ -d $old_home && ! -L $old_home ]]; then
      warn "'$old_home' does not exist or is not a folder, nothing to migrate"
      exit 0
    fi
    ```

Loving the Steam Deck so far!

**Comment by bertogg** [comment]
> steamos-finish-oobe-migration.service is incorrectly failing. The bash script /usr/lib/steamos/steamos-finish-oobe-migration has this line:

> `# Shouldn't have been invoked if $old_home isn't here or is a symlink.`
> `[[ -d $old_home && ! -L $old_home ]] || die "'$old_home' does not exist or is not a folder, nothing to migrate"`

What's wrong with this? It works fine for me here.

**Comment by docwhat** [comment]
It shouldn't be using `die` which exits with a non-zero exit code and causes `systemctl --failed` to show that service as failing.

If the `$old_home` is a symlink to a directory then that means the script has run at some point in the past and completed successfully or it never needed migration.  There is no error, this is a good thing.

The correct response is to log a `warn` or an `info` and exit with an exit code of `0` to tell systemd that the service is happy.

**Comment by bertogg** [comment]
Ah, I see, I think that the problem is that the service shouldn't run in the first place.

`ConditionPathIsDirectory=/home/doorstop` is not enough because this check follows symbolic links. Because of that the `doorstop -> deck` symlink doesn't prevent the service from running, and it fails as you point out.

Thanks!

**Comment by bertogg** [comment]
Yes, that fix is correct :slightly_smiling_face: 

Just to clarify, these issues are being handled and will be fixed in a future release.

**Comment by bertogg** [comment]
These services should be working fine now. We can close this issue.

**Comment by Sunspark-007** [comment]
@bertogg When were they fixed? I'm in 3.4 and "grub-recordfail.service" is listed as failed.

**Comment by bertogg** [comment]
Are you still in 3.4 ?

This was fixed a long time ago (for new installations). If come from an early version you might still have the grub-recordfail service lying around. `systemctl disable` is indeed a correct way to solve the issue.
